types/json/binary_functions.go

```diff
@@ -781,6 +781,96 @@ func CompareBinary(left, right BinaryJSON) int {
 	return cmp
 }
 
+// MergePatchBinary implements RFC7396
+// https://datatracker.ietf.org/doc/html/rfc7396
+func MergePatchBinary(bjs []*BinaryJSON) (*BinaryJSON, error) {
+	var err error
+	length := len(bjs)
+
+	// according to the implements of RFC7396
+	// when the last item is not object
+	// we can return the last item directly
+	for i := length - 1; i >= 0; i-- {
+		if bjs[i] == nil || bjs[i].TypeCode != TypeCodeObject {
+			bjs = bjs[i:]
+			break
+		}
+	}
+
+	target := bjs[0]
+	for _, patch := range bjs[1:] {
+		target, err = mergePatchBinary(target, patch)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return target, nil
+}
+
+func mergePatchBinary(target, patch *BinaryJSON) (result *BinaryJSON, err error) {
+	if patch == nil {
+		return nil, nil
+	}
+
+	if patch.TypeCode == TypeCodeObject {
+		if target == nil {
+			return nil, nil
+		}
+
+		keyValMap := make(map[string]BinaryJSON)
+		if target.TypeCode == TypeCodeObject {
+			elemCount := target.GetElemCount()
+			for i := 0; i < elemCount; i++ {
+				key := target.objectGetKey(i)
+				val := target.objectGetVal(i)
+				keyValMap[string(key)] = val
+			}
+		}
+		var tmp *BinaryJSON
+		elemCount := patch.GetElemCount()
+		for i := 0; i < elemCount; i++ {
+			key := patch.objectGetKey(i)
+			val := patch.objectGetVal(i)
+			k := string(key)
+
+			targetKV, exists := keyValMap[k]
+			if val.TypeCode == TypeCodeLiteral && val.Value[0] == LiteralNil {
+				if exists {
+					delete(keyValMap, k)
+				}
+			} else {
+				tmp, err = mergePatchBinary(&targetKV, &val)
+				if err != nil {
+					return result, err
+				}
+
+				keyValMap[k] = *tmp
+			}
+		}
+
+		length := len(keyValMap)
+		keys := make([][]byte, 0, length)
+		for key := range keyValMap {
+			keys = append(keys, []byte(key))
+		}
+		sort.Slice(keys, func(i, j int) bool {
+			return bytes.Compare(keys[i], keys[j]) < 0
+		})
+		length = len(keys)
+		values := make([]BinaryJSON, 0, len(keys))
+		for i := 0; i < length; i++ {
+			values = append(values, keyValMap[string(keys[i])])
+		}
+
+		binaryObject, e := buildBinaryObject(keys, values)
+		if e != nil {
+			panic("mergePatchBinary should never panic, please contact the TiDB team for help")
```

Comment:
Why not return the error?

Reply:
I copy from mergePatchBinary.
![image](https://user-images.githubusercontent.com/6133860/125047707-d0c09200-e0d1-11eb-9c99-98cc8ea7e0ba.png)

Reply:
May be error should be return in here

Reply:
I think return error is better, although the error is always nil.